src/Microsoft.TestPlatform.Client/TestPlatform.cs

```diff
@@ -303,21 +277,133 @@ private void AddExtensionAssembliesFromSource(IEnumerable<string> sources)
     /// </summary>
     private static void AddExtensionAssembliesFromExtensionDirectory()
     {
+        // This method runs before adapter initialization path, ideally we should replace this mechanism
+        // this is currently required because we need TestHostProvider to be able to resolve.
+        var runSettings = RunSettingsManager.Instance.ActiveRunSettings.SettingsXml;
+        var runConfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(runSettings);
+        var strategy = runConfiguration.TestAdapterLoadingStrategy;
+
         var fileHelper = new FileHelper();
-        var extensionsFolder = Path.Combine(
-            Path.GetDirectoryName(
-                typeof(TestPlatform).GetTypeInfo().Assembly.GetAssemblyLocation()),
-            "Extensions");
+        var defaultExtensionPaths = Enumerable.Empty<string>();
 
+        // Explicit adapter loading
+        if ((strategy & TestAdapterLoadingStrategy.Explicit) == TestAdapterLoadingStrategy.Explicit)
+        {
+            defaultExtensionPaths = RunSettingsUtilities.GetTestAdaptersPaths(runSettings)
+                .SelectMany(path => ExpandTestAdapterPaths(path, fileHelper, strategy))
+                .Union(defaultExtensionPaths);
+        }
+
+        var extensionsFolder = Path.Combine(Path.GetDirectoryName(typeof(TestPlatform).GetTypeInfo().Assembly.GetAssemblyLocation()), "Extensions");
         if (fileHelper.DirectoryExists(extensionsFolder))
         {
-            var defaultExtensionPaths = fileHelper.EnumerateFiles(
-                extensionsFolder,
-                SearchOption.TopDirectoryOnly,
-                ".dll",
-                ".exe");
+            // Load default runtime providers
+            if ((strategy & TestAdapterLoadingStrategy.DefaultRuntimeProviders) == TestAdapterLoadingStrategy.DefaultRuntimeProviders)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, TestPlatformConstants.RunTimeEndsWithPattern)
+                    .Union(defaultExtensionPaths);
+            }
+
+            // Default extension loader
+            if (strategy == TestAdapterLoadingStrategy.Default
+                || (strategy & TestAdapterLoadingStrategy.ExtensionsDirectory) == TestAdapterLoadingStrategy.ExtensionsDirectory)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, ".dll", ".exe")
+                    .Union(defaultExtensionPaths);
+            }
+        }
+
+        TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths.Distinct();
```

Comment:
Is it really worth having the lazy behavior?

Reply:
Probably no, but it's the default behavior here, I'll hopefully remove this code path entirely later on.